[CodeComplete] Default parameter names of completed closure to internal names

ahoppen opened the pull request:
If have a function that takes a trailing closure as follows
```swift
func sort(callback: (_ left: Int, _ right: Int) -> Bool) {}
```
completing a call to `sort` and expanding the trailing closure results in
```swift
sort { <#Int#>, <#Int#> in
  <#code#>
}
```

We should be doing a better job here and defaulting the trailing closure's to the internal names specified in the function signature. I.e. the final result should be
```swift
sort { left, right in
  <#code#>
}
```

This commit does exactly that.

Firstly, it keeps track of the closure's internal names (as specified in the declaration of `sort`) in the closure's type through a new `InternalLabel` property in `AnyFunctionType::Param`. Once the type containing the parameter gets canonicalized, the internal label is dropped.

Secondly, it adds a new option to `ASTPrinter` to always try and print parameter labels. With this option set to true, it will always print external paramter labels and, if they are present, print the internal parameter label as `_ <internalLabel>`.

Finally, we can use this new printing mode to print the trailing closure’s type as
```
<#T##callback: (Int, Int) -> Bool##(_ left: Int, _ right: Int) -> Bool#>
```

This is already correctly expanded by code-expand to the desired result. I also added a test case for that behaviour.

rintaro commented:
I guess, at least, you need to change `FunctionType::Profile()` and `AnyFunctionType::Param::operator==` to take `InternalLabel` into account. Also `getCanonicalType()` should return function types without any `InternalLabel` so that type checker can know `(_ a: Int) -> Void` and `(_ b: Int) -> Void` is equivalent.

ahoppen commented:
OK, I have done a 180º turnaround. After sleeping over it, I figured the internal names really have nothing to do in the type system because they don’t describe a function’s type. Instead, I discovered that we still keep the internal names around in the parameters `TypeRepr` and can extract them from there. The extraction is a little bit tedious, but I think it’s better than cluttering the type system with information that only code completion cares about. What do you think?

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
OK, 180º turnaround again to the way I was facing before. Turns out my initial implementation was nearly there and I was just confused because a whole bunch of the test suite failed on my machine, but I didn’t realize that was because it was already failing for my local setup on `main`.

So, internal names are again part of the type system and are cleared when canonicalizing the type.

@swift-ci Please test

ahoppen commented:
Test failure on Linux seems unrelated. Let’s go for smoke tests only.

@swift-ci Please smoke test Linux

ahoppen commented:
@swift-ci Please smoke test Linux

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented on include/swift/AST/Types.h:
```diff
@@ -2833,21 +2845,29 @@ class AnyFunctionType : public TypeBase {
 
     bool operator==(Param const &b) const {
       return (Label == b.Label &&
+              InternalLabel == b.InternalLabel &&
```
It would definitely work, but I thought it’s unintuitive if the `==` operator silently ignores one of the properties altogether, that’s why I went with the dedicated method. 

Thinking about this some more, I think we should 
a) make `==` check for strict equality without canonicalizing types (or remove it entirely) and provide `equals` functions that check the various forms of equality a `Param` can take or
b) accept that `==` that the definition of `==` is a bit sloppy, embrace it, don’t check for `InternalLabel` and add a doc comment describing the catches.

I can see advantages in either approach (clearer semantics in (a) but more concise syntax in (b)). I think I tend to lean towards (a) but don’t have too strong feelings about it. Do you @rintaro?

rintaro replied:
I'd suggest (b) in this PR. Here's the reasoning: Considering `Type::operator==` is `delete` we probably should delete `AnyFunctionType::Param::operator==` and make `isEqual` which compares "canonical" forms of the values. But, that also applies to `AnyFunctionType::Yield::operator==`. I don't think we should do it in this PR because `AnyFunctionType::Yield` is unrelated. Also we should not change the semantic meaning of `AnyFunctionType::Param::operator==` in this PR.

ahoppen replied:
Sounds good to me.

ahoppen replied:
After looking into it some more, I agree that ignoring the internal name in `operator==` is *absolutely* the right choice. I wasn’t aware that we are *always* using canonical types in `operator==` for types and that it’s rather more difficult to compare the sugared types. As such, ignoring the internal name in `Param::operator==` follows the current paradigm.

ahoppen committed:
[CodeComplete] Default parameter names of completed closure to internal names

If have a function that takes a trailing closure as follows
```
func sort(callback: (_ left: Int, _ right: Int) -> Bool) {}
```
completing a call to `sort` and expanding the trailing closure results in
```
sort { <#Int#>, <#Int#> in
  <#code#>
}
```

We should be doing a better job here and defaulting the trailing closure's to the internal names specified in the function signature. I.e. the final result should be
```
sort { left, right in
  <#code#>
}
```

This commit does exactly that.

Firstly, it keeps track of the closure's internal names (as specified in the declaration of `sort`) in the closure's type through a new `InternalLabel` property in `AnyFunctionType::Param`. Once the type containing the parameter gets canonicalized, the internal label is dropped.

Secondly, it adds a new option to `ASTPrinter` to always try and print parameter labels. With this option set to true, it will always print external paramter labels and, if they are present, print the internal parameter label as `_ <internalLabel>`.

Finally, we can use this new printing mode to print the trailing closure’s type as
```
<#T##callback: (Int, Int) -> Bool##(_ left: Int, _ right: Int) -> Bool#>
```

This is already correctly expanded by code-expand to the desired result. I also added a test case for that behaviour.

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please smoke test Linux

ahoppen commented:
@swift-ci Please smoke test Windows

rintaro reviewed (commented):
This looks good to me. We still need serialization support for this. Do you intend to do it in followups?

ahoppen commented on lib/AST/ASTContext.cpp:
```diff
@@ -3090,6 +3090,10 @@ isFunctionTypeCanonical(ArrayRef<AnyFunctionType::Param> params,
   for (auto param : params) {
     if (!param.getPlainType()->isCanonical())
       return false;
+    if (!param.getInternalLabel().empty()) {
+      // Canonical types don't have internal labels
+      return false;
+    }
```
Ah, I didn't realize there was `isGenericFunctionTypeCanonical`. Added it there as well and renamed `isFunctionTypeCanonical` to `isAnyFunctionTypeCanonical` (with `Any`) to imply there is a generic version as well.

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
> This looks good to me. We still need serialization support for this. Do you intend to do it in followups?

I’ll look into it in a follow-up PR.

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please smoke test Windows

ahoppen commented:
Windows failures is unrelated. Merging.

rintaro commented on test/IDE/complete_call_arg.swift:
```diff
@@ -120,6 +120,12 @@
 
 // RUN: %target-swift-ide-test -code-completion -source-filename %s -code-completion-token=NAMED_PARAMETER_WITH_LEADING_VARIADIC | %FileCheck %s -check-prefix=NAMED_PARAMETER_WITH_LEADING_VARIADIC
 
+// RUN: %target-swift-ide-test -code-completion -source-filename %s -code-completion-token=CLOSURE_PARAM_WITH_INTERNAL_NAME | %FileCheck %s -check-prefix=CLOSURE_PARAM_WITH_INTERNAL_NAME
+// RUN: %target-swift-ide-test -code-completion -source-filename %s -code-completion-token=CLOSURE_PARAM_WITH_PARENS | %FileCheck %s -check-prefix=CLOSURE_PARAM_WITH_PARENS
+// RUN: %target-swift-ide-test -code-completion -source-filename %s -code-completion-token=OPTIONAL_CLOSURE_PARAM | %FileCheck %s -check-prefix=OPTIONAL_CLOSURE_PARAM
+// RUN: %target-swift-ide-test -code-completion -source-filename %s -code-completion-token=ESCAPING_OPTIONAL_CLOSURE_PARAM | %FileCheck %s -check-prefix=ESCAPING_OPTIONAL_CLOSURE_PARAM
+// RUN: %target-swift-ide-test -code-completion -source-filename %s -code-completion-token=COMPLETE_CLOSURE_PARAM_WITHOUT_INTERNAL_NAMES | %FileCheck %s -check-prefix=COMPLETE_CLOSURE_PARAM_WITHOUT_INTERNAL_NAMES
+
```
Ah sorry, could you remove these lines? They are redundant. Since https://github.com/apple/swift/pull/36725 , this test file uses `-batch-code-completion` mode which checks all tokens in a single invocation. See also https://github.com/apple/swift/pull/28946

ahoppen replied:
Didn’t notice this when rebasing. Thanks. I’ve set up a PR here: https://github.com/apple/swift/pull/36796
